src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11251,14 +11251,45 @@ private bool ScanCast(bool forPattern = false)
                     // following a cast.
                     return !forPattern || this.CurrentToken.Kind switch
                     {
-                        SyntaxKind.PlusToken => true,
-                        SyntaxKind.MinusToken => true,
-                        SyntaxKind.AmpersandToken => true,
-                        SyntaxKind.AsteriskToken => true,
+                        SyntaxKind.PlusToken or
+                        SyntaxKind.MinusToken or
+                        SyntaxKind.AmpersandToken or
+                        SyntaxKind.AsteriskToken or
                         SyntaxKind.DotDotToken => true,
-                        _ => CanFollowCast(this.CurrentToken.Kind)
+                        SyntaxKind.IdentifierToken => !isBinaryPattern(),
+                        var tk => CanFollowCast(tk)
                     };
 
+                    bool isBinaryPattern()
+                    {
+                        if (!isBinaryPatternKeyword())
+                        {
+                            return false;
+                        }
+
+                        bool lastTokenIsBinaryOperator = true;
+
+                        EatToken();
+                        while (isBinaryPatternKeyword())
+                        {
+                            // If the parenthesized type is followed by an even-number of binary pattern tokens, it is a cast.
```

Comment:
>If the parenthesized type is followed by an even-number of binary pattern tokens, it is a cast. [](start = 31, length = 95)

This seems a strange rule to specify and adopt. I think we should just take a precedence (for example, prefer the `or` contextual keyword to be interpreted as a binary pattern.

Reply:
We already accept or/and as identifiers in cases not involving casts (added tests for those).

This is a consequence of the grammar not a "rule". Maybe a different wording would help?

Reply:
Actually, considering the check at the end, this comment is not correct.